packages/babel-plugin-proposal-pipeline-operator/src/index.js

```diff
@@ -1,64 +1,23 @@
 import { declare } from "@babel/helper-plugin-utils";
 import syntaxPipelineOperator from "@babel/plugin-syntax-pipeline-operator";
-import { types as t } from "@babel/core";
+import minimalVisitor from "./minimalVisitor";
+import smartVisitor from "./smartVisitor";
 
-export default declare(api => {
+export default declare((api, options) => {
   api.assertVersion(7);
 
+  // FIXME: Hacky... Need a way to get rid of this
+  delete smartVisitor._exploded;
+  delete smartVisitor._verified;
+
+  const visitorsPerProposal = {
+    minimal: minimalVisitor,
+    smart: smartVisitor,
+  };
+
   return {
     name: "proposal-pipeline-operator",
     inherits: syntaxPipelineOperator,
-
-    visitor: {
-      BinaryExpression(path) {
-        const { scope } = path;
-        const { node } = path;
-        const { operator, left } = node;
-        let { right } = node;
-        if (operator !== "|>") return;
-
-        let optimizeArrow =
-          t.isArrowFunctionExpression(right) &&
-          t.isExpression(right.body) &&
-          !right.async &&
-          !right.generator;
-        let param;
-
-        if (optimizeArrow) {
-          const { params } = right;
-          if (params.length === 1 && t.isIdentifier(params[0])) {
-            param = params[0];
-          } else if (params.length > 0) {
-            optimizeArrow = false;
-          }
-        } else if (t.isIdentifier(right, { name: "eval" })) {
-          right = t.sequenceExpression([t.numericLiteral(0), right]);
-        }
-
-        if (optimizeArrow && !param) {
-          // Arrow function with 0 arguments
-          path.replaceWith(t.sequenceExpression([left, right.body]));
-          return;
-        }
-
-        const placeholder = scope.generateUidIdentifierBasedOnNode(
-          param || left,
-        );
-        scope.push({ id: placeholder });
-        if (param) {
-          path.get("right").scope.rename(param.name, placeholder.name);
-        }
-
-        const call = optimizeArrow
-          ? right.body
-          : t.callExpression(right, [t.cloneNode(placeholder)]);
-        path.replaceWith(
-          t.sequenceExpression([
-            t.assignmentExpression("=", t.cloneNode(placeholder), left),
-            call,
-          ]),
-        );
-      },
-    },
+    visitor: visitorsPerProposal[options.proposal],
```

Comment:
We should add validation for the `proposal` optoin.

Reply:
I think the syntax plugin already does this.

Reply:
Do we need to add it here too?

Reply:
Is it possible for this change to be observable by a test?

Reply:
Yes, you can test the error message by adding `"error": "Expected error message"` in the options.json file of the fixture.

Reply:
Well, even then: That test still wouldn’t distinguish between the parser throwing the error versus the transform plugin throwing the error, right?

Reply:
Yeah that's right. But usually parser error messages are different from plugin error messages. That is the important part; not where the error actually comes from.

Reply:
I find the syntax plugin message (below) is clear enough in this case.

```
The pipeline operator plugin requires a 'proposal' option.
'proposal' must be one of: minimal, smart. More details:
https://babeljs.io/docs/en/next/babel-plugin-proposal-pipeline-operator
```

Maybe it should be moved to the transform plugin. What about holding it off for a later PR?

Reply:
Oh I didn't noticed that the link alreay points to the transform plugin. It is ok as-is :+1: